ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Server.scala

```diff
@@ -0,0 +1,395 @@
+/*
+ * Copyright 2019 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.ember.core.h2
+
+import cats._
+import cats.effect._
+import cats.effect.syntax.all._
+import cats.syntax.all._
+import fs2._
+import fs2.io.net._
+import org.http4s._
+import org.typelevel.ci._
+import org.typelevel.log4cats.Logger
+import scodec.bits._
+
+import scala.concurrent.duration._
+
+import H2Frame.Settings.ConnectionSettings.{default => defaultSettings}
+
+private[ember] object H2Server {
+
+  /*
+  3 Mechanism into H2
+
+  TlsContext => Yes => ALPN =>  h2       => HTTP2
+                                http/1.1 => HTTP1
+                                Nothing  => Http1
+                No  =>                      Http1
+
+  This is the tricky one, must read the initial bytes
+  if they are the prelude then upgrade, but only on when
+  the connection is first established
+
+  Client Prelude Bytes                   => Http2 // Http2-prior-kno
+  checkConnectionPreface
+  Note: implementations that support HTTP/2 over TLS MUST use protocol
+    negotiation in TLS.
+    So if TLS is used then this method is not allowed.
+
+
+  H2c
+            Request
+            Connection: Upgrade, HTTP2-Settings
+            Upgrade: h2c
+            HTTP2-Settings: <base64url encoding of HTTP/2 SETTINGS payload>
+              =>
+                HTTP/1.1 101 Switching Protocols
+                Connection: Upgrade
+                Upgrade: h2c
+
+                Socket                    => Http2
+
+            Normal                        => Resp
+
+   */
+
+  private val upgradeResponse: Response[fs2.Pure] = Response(
+    status = Status.SwitchingProtocols,
+    httpVersion = HttpVersion.`HTTP/1.1`,
+    headers = Headers(
+      "connection" -> "Upgrade",
+      "upgrade" -> "h2c",
+    ),
+  )
+  // Apply this, if it ever becomes Some, then rather than the next request, become an h2 connection
+  def h2cUpgradeHttpRoute[F[_]: Concurrent]: HttpRoutes[F] =
+    cats.data.Kleisli[({ type L[A] = cats.data.OptionT[F, A] })#L, Request[F], Response[F]] {
```

Comment:
Any reason this one isn't kind projected?

Reply:
You can't kind project this for scala 3. It sucks.